extension/src/paymentHandler/cancel-or-refund.handler.js

```diff
@@ -1,51 +1,63 @@
+const {
+  CTP_INTERACTION_TYPE_MAKE_PAYMENT,
+  CTP_INTERACTION_TYPE_CANCEL_OR_REFUND,
+  CTP_TXN_STATE_PENDING,
+} = require('../config/constants')
 
-async function handlePayment (paymentObject) {
+async function execute (paymentObject) {
   const { request, response } = await _cancelOrRefundPayment(paymentObject)
-  const status = response.status === 200 ? c.SUCCESS : c.FAILURE
-  const responseBody = await response.json()
+
+  const addInterfaceInteractionAction = pU.createAddInterfaceInteractionAction({
+      request, response, type: CTP_INTERACTION_TYPE_CANCEL_OR_REFUND
+  })
+
+  if (isEmpty(response.pspReference))
+    return { actions: [ addInterfaceInteractionAction ] }
+
   const actions = [
-    pU.ensureAddInterfaceInteractionAction({
-      paymentObject, request, response: responseBody, type: c.CTP_INTERACTION_TYPE_CANCEL_REFUND, status
-    })
+    addInterfaceInteractionAction,
+    ...getTransactionActions(paymentObject, response.pspReference)
   ]
-  if (status === c.SUCCESS) {
-    const cancelTransaction = pU.getCancelAuthorizationTransactionInit(paymentObject)
-    const refundTransaction = pU.getRefundTransactionInit(paymentObject)
-    if (cancelTransaction)
-      actions.push(
-        pU.createChangeTransactionStateAction(cancelTransaction.id, c.CTP_TXN_STATE_PENDING)
-      )
-    else if (refundTransaction)
-      actions.push(
-        pU.createChangeTransactionStateAction(refundTransaction.id, c.CTP_TXN_STATE_PENDING)
-      )
-  }
+
   return { actions }
 }
 
 async function _cancelOrRefundPayment (paymentObject) {
-  const transaction = pU.getAuthorizationTransactionSuccess(paymentObject)
+  const authorizationTransaction = pU.getAuthorizationTransactionSuccess(paymentObject)
   // "originalReference: The original pspReference of the payment that you want to cancel or refund.
   // This reference is returned in the response to your payment request, and in the AUTHORISATION notification."
-  const body = {
-    merchantAccount: config.adyen.merchantAccount,
-    originalReference: transaction.interactionId,
-    reference: paymentObject.custom.fields.merchantReference
-  }
+  const cancelOrRefundRequestObj = { originalReference: authorizationTransaction.interactionId }
 
-  const request = {
-    method: 'POST',
-    body: JSON.stringify(body),
-    headers: { 'x-api-key': config.adyen.apiKey, 'Content-Type': 'application/json' }
+  const interfaceInteraction = paymentObject.interfaceInteractions
+    .find(interaction => interaction.fields.type === CTP_INTERACTION_TYPE_MAKE_PAYMENT)
+  if (interfaceInteraction && !isEmpty(interfaceInteraction.fields.response)) {
+    // OPTIONAL:
+    // Specifies a unique identifier for payment modification.
+    // The reference field is useful to tag a partial cancel or refund for future reconciliation.
+    const makePaymentResponse = JSON.parse(interfaceInteraction.fields.response)
+    cancelOrRefundRequestObj.reference = makePaymentResponse.merchantReference
```

Comment:
Why do you put `merchantReference` for this field? There's nothing about this part in the documentation. Where did you find it?

Reply:
General documentation is here: https://docs.adyen.com/checkout/cancel-or-refund#cancel-or-refund-a-payment 

a reference that is passed with /payment request: https://docs.adyen.com/api-explorer/#/PaymentSetupAndVerificationService/v52/post/payments__reqParam_reference 
will return in response as a merchant reference.

Check here for an example: https://github.com/commercetools/commercetools-adyen-integration/blob/01d482cc780f0bc23bbc5cd0ecc58abade49d5c1/extension/test/unit/cancel-or-refund.handler.spec.js#L29-L32

As described in the comments, it's an optional field. Documentations do not say merchant reference because it could be any reference, so merchant reference could be a reference field for our scenario. Before web-components that field was a custom field passed by the front end. Now we are using the `makePayment` interaction to get the reference.

After that long comment, if you're asking a different thing, please give an example of your question.

Reply:
No, that's what I was asking for :) Theoretically there could be multiple `interfaceInteraction` for `makePayment`, so I guess you should get the newest one. Is it better if you take the response from the custom field `makePaymentResponse`?

Reply:
> Theoretically there could be multiple interfaceInteraction for makePayment, so I guess you should get the newest one.

This is a good point 👍

Reply:
The issue is created for the additional payment details : 
https://github.com/commercetools/commercetools-adyen-integration/issues/289

Refactoring for the comment: https://github.com/commercetools/commercetools-adyen-integration/pull/286/commits/d0baa3b9256518b24b8ced13c973d8e3c099f24a